File: java/sort/array/SelectionSort.java
package sort.array;

public class SelectionSort {
    public void sort(int[] nums) {
        int start = 0;
        int end = nums.length - 1;
        while (start <= end) {
            int maxIndex = start;
            int minIndex = start;
            for (int i = start + 1; i <= end; ++i) {
                if (nums[maxIndex] < nums[i]) {
                    maxIndex = i;
                } else if (nums[minIndex] > nums[i]) {
                    minIndex = i;
                }
            }

            swap(nums, minIndex, start);
            if (maxIndex == start) {
                maxIndex = minIndex;
            }
            swap(nums, maxIndex, end);

            ++start;
            --end;
        }
    }

    private void swap(int[] nums, int i, int j) {
        if (i != j) {
            nums[i] ^= nums[j];
            nums[j] ^= nums[i];
            nums[i] ^= nums[j];
        }
    }
}

File: py/sort/array/SelectionSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class SelectionSort:
    def sort(self, nums: List[int]) -> None:
        start, end = 0, len(nums) - 1
        while start < end:
            maxIndex, minIndex = start, start
            for i, value in enumerate(nums[start : end + 1]):
                if nums[maxIndex] < value:
                    maxIndex = i + start
                elif nums[minIndex] > value:
                    minIndex = i + start

            nums[minIndex], nums[start] = nums[start], nums[minIndex]
            if maxIndex == start:
                maxIndex = minIndex
            nums[maxIndex], nums[end] = nums[end], nums[maxIndex]

            start += 1
            end -= 1
